README
# crt - the C raytracer #

Raytracing is just one technique used in generating realistic images using a
computer. In the real world, what we see is a direct result of light photons
being emitted from a source (such as the sun, a light bulb, etc), bouncing off
objects, and eventually being collected by our eyes. As simple as this is, it's
still a lot of work for a computer to do.

This is because many of the photons aren't actually percieved by us. Raytracing
greatly simplifies the way lighting works by viewing light in the opposition
direction. Instead of tracing photons from the sources, we trace a ray from our
eye through the scene. This means we only calculate the rays of light that
matter to us.

The raytracing algorithm itself is relatively simple. For each pixel in the
image, we trace a ray through that pixel and into the scene. We then check for
an intersection with any of the objects in the scene. If the ray hits nothing,
then that pixel becomes black. If the ray does intersect an object, that pixel
becomes that objects color.

After that, the pixel's color can be modified based on various other algorithms,
such as lighting algorithms, reflections, etc. Reflections can easily be
calculated by using the ray tracing algorithm recursively. Essentially, the
reflective object becomes the "eye", and the scene is re-traced from that view
point.

## scene files ##

crt uses a modular scene input scheme to allow scenes to be loaded from a
variety of sources. Currently crt uses an XML plugin to load scene definition
files from an XML file. An example of an object's XML:

    <sphere x="1.0" y="2.0" z="0.0" radius="1.0">
     <material r="1.0" g="0.0" b="0.0" diffuse="1.0" />
    </sphere>

## image output ##

Again, crt uses a modular image output scheme to write images. Currently crt
supports writing directly to a bitmap image.

## features ##

Currently crt features the following:

  * 3 different shapes: spheres, axis-aligned boxes, and planes
  * Point and Area light sources
  * Shadows
  * Anti-Aliasing
  * Reflections
  * Refractions

## generated images ##

<div class="image">
 <img alt="scene1" src="/code/crt/images/scene1.png" />
 <div class="image_footer">Scene 1 (<a href="/code/crt/images/scene1_1920x1200.png">1920x1200</a>)</div>
</div>

<div class="image">
 <img alt="scene2" src="/code/crt/images/scene2.png" />
 <div class="image_footer">Scene 2 (<a href="/code/crt/images/scene2_1920x1200.png">1920x1200</a>)</div>
</div>

<div class="image">
 <img alt="scene3" src="/code/crt/images/scene3.png" />
 <div class="image_footer">Scene 3 (<a href="/code/crt/images/scene3_1920x1200.png">1920x1200</a>)</div>
</div>

## GIT ##

You can checkout a bleeding-edge version of crt with git using the following
command:

    git clone http://voidsplat.org/code/crt/crt.git


CHANGELOG
CRT Changelog

* 0.2.0:
    * Added soft shadows, set the number of shadow calculations via
      the `--shadows` command line argument. For best results, the value
      should be a multiple of 64
    * Fixed a few memory leaks

* 0.1.0:
    * Initial Release


crt.c
#include <stdio.h>
#include <stdlib.h>
#include <dlfcn.h>
#include <math.h>
#include <string.h>

#include "vector.h"
#include "objects.h"

#define EPSILON     0.001f
#define TRACE_DEPTH 6

void render( int, int, color***, primitive*, int, int );
color* trace( ray*, primitive*, int, float, float*, int );
intersection* intersect( ray*, primitive* );
float calc_shade( primitive*, vector*, vector*, vector*, primitive*, int );
void usage(void);

void render( int width, int height, color ***image, primitive *scene,
        int aa_level, int shadows ) {
    float world_left, world_right, world_top, world_bot;
    float delta_x, delta_y;
    float screen_x, screen_y;
    float tmpf;
    int x, y;
    int aa_x, aa_y;
    int aa_root;
    vector origin, dir;
    ray r;
    color *tmpc;

    world_left = -(4.0f * ((float)width / (float)height) );
    world_right = -world_left;
    world_top = 4.0f;
    world_bot = -4.0f;

    /* The amount to shift for each pixel */
    delta_x = (world_right - world_left) / width;
    delta_y = (world_bot - world_top) / height;

    origin.x = 0.0f;
    origin.y = 0.0f;
    origin.z = -5.0f;

    aa_root = (int)sqrt( aa_level );

    screen_y = world_top;
    for( y = 0; y < height; y++ ) {
        screen_x = world_left;
        for( x = 0; x < width; x++ ) {
            for( aa_x = 0; aa_x < aa_root; aa_x++ ) {
                for( aa_y = 0; aa_y < aa_root; aa_y++ ) {
                    dir.x = screen_x + delta_x * aa_x / (float)aa_root;
                    dir.y = screen_y + delta_y * aa_y / (float)aa_root;
                    dir.z = 0.0f;

                    vect_sub( &dir, &origin );
                    vect_normalize( &dir );

                    r.origin = &origin;
                    r.dir = &dir;

                    tmpc = trace( &r, scene, 0, 1.0f, &tmpf, shadows );
                    if( image[y][x] == NULL ) {
                        image[y][x] = tmpc;
                    }
                    else {
                        vect_add( image[y][x], tmpc );
                        free( tmpc );
                    }
                }
            }

            vect_multf( image[y][x], 1.0f / aa_level );

            screen_x += delta_x;
        }
        screen_y += delta_y;
    }
}

color* trace( ray *aray, primitive *scene, int depth, float refr, float *dist,
        int shadows ){
    intersection *isect;
    primitive *prim, *iter;
    vector isect_pt, pn, lv, ln, tmpv1, tmpv2;
    ray tmpr;
    float tmpf1, tmpf2, tmpf3, shade;
    color *tmpc;
    color *c = (color*)malloc( sizeof( color ) );
    if( c == NULL ) {
        fprintf( stderr, "*** error: could not allocate color memory\n" );
        exit( 1 );
    }

    c->x = 0.0f;
    c->y = 0.0f;
    c->z = 0.0f;

    isect = intersect( aray, scene );

    if( isect == NULL ) {
        return c;
    }

    *dist = isect->dist;

    prim = isect->prim;

    if( prim->is_light ) {
        vect_copy( c, &(isect->prim->mat.col) );
        free( isect );
        return c;
    }

    vect_copy( &isect_pt, aray->dir );
    vect_multf( &isect_pt, isect->dist );
    vect_add( &isect_pt, aray->origin );

    prim->normal( prim, &isect_pt, &pn );

    iter = scene;
    while( iter != NULL ) {
        if( iter->is_light ) {
            vect_copy( &lv, &iter->center );
            vect_sub( &lv, &isect_pt );

            vect_copy( &ln, &lv );
            vect_normalize( &ln );

            shade = calc_shade( iter, &isect_pt, &lv, &ln, scene, shadows );

            if( shade > 0.0f ) {
                /* determine the diffuse component */
                tmpf1 = prim->mat.diffuse;
                if( tmpf1 > 0.0f ) {
                    tmpf2 = vect_dot( &pn, &ln );
                    if( tmpf2 > 0.0f ) {
                        tmpf1 *= tmpf2 * shade;
                        vect_copy( &tmpv1, &prim->mat.col );
                        vect_mult( &tmpv1, &iter->mat.col );
                        vect_multf( &tmpv1, tmpf1 );
                        vect_add( c, &tmpv1 );
                    }
                }

                /* determine the specular component */
                tmpf1 = prim->mat.specular;
                if( tmpf1 > 0.0f ) {
                    vect_copy( &tmpv1, &pn );
                    vect_copy( &tmpv2, &ln );
                    tmpf2 = 2.0f * vect_dot( &ln, &pn );
                    vect_multf( &tmpv1, tmpf2 );
                    vect_sub( &tmpv2, &tmpv1 );
                    tmpf2 = vect_dot( aray->dir, &tmpv2 );
                    if( tmpf2 > 0.0f ) {
                        tmpf1 = powf( tmpf2, 20.0f ) * tmpf1 * shade;
                        vect_copy( &tmpv1, &iter->mat.col );
                        vect_multf( &tmpv1, tmpf1 );
                        vect_add( c, &tmpv1 );
                    }
                }
            }
        }

        iter = iter->next;
    }

    /* calculate reflection */
    if( prim->mat.refl > 0.0f && depth < TRACE_DEPTH ) {
        vect_copy( &tmpv1, &pn );
        vect_multf( &tmpv1, 2.0f * vect_dot( &pn, aray->dir ) );
        vect_copy( &tmpv2, aray->dir );
        vect_sub( &tmpv2, &tmpv1 );

        vect_copy( &tmpv1, &tmpv2 );
        vect_multf( &tmpv1, EPSILON );
        vect_add( &tmpv1, &isect_pt );

        tmpr.origin = &tmpv1;
        tmpr.dir = &tmpv2;

        tmpc = trace( &tmpr, scene, depth + 1, refr, &tmpf1, shadows );
        vect_multf( tmpc, prim->mat.refl );

        vect_copy( &tmpv1, &prim->mat.col );
        vect_mult( &tmpv1, tmpc );
        vect_add( c, &tmpv1 );

        free( tmpc );
    }

    /* calculate refraction */
    if( prim->mat.is_refr && depth < TRACE_DEPTH ) {
        vect_copy( &tmpv1, &pn );
        if( isect->inside ) {
            vect_multf( &tmpv1, -1.0f );
        }

        tmpf1 = refr / prim->mat.refr;
        tmpf2 = -( vect_dot( &tmpv1, aray->dir ) );
        tmpf3 = 1.0f - tmpf1 * tmpf1 * (1.0f - tmpf2 * tmpf2);
        if( tmpf3 > 0.0f ) {
            vect_copy( &tmpv2, aray->dir );
            vect_multf( &tmpv2, tmpf1 );
            vect_multf( &tmpv1, tmpf1 * tmpf2 - sqrtf( tmpf3 ) );
            vect_add( &tmpv1, &tmpv2 );

            vect_copy( &tmpv2, &tmpv1 );
            vect_multf( &tmpv2, EPSILON );
            vect_add( &tmpv2, &isect_pt );

            tmpr.origin = &tmpv2;
            tmpr.dir = &tmpv1;

            tmpc = trace( &tmpr, scene, depth + 1, refr, &tmpf1, shadows );

            vect_copy( &tmpv1, &prim->mat.col );
            vect_multf( &tmpv1, prim->mat.absorb * tmpf1 );

            tmpv2.x = expf( -tmpv1.x );
            tmpv2.y = expf( -tmpv1.y );
            tmpv2.z = expf( -tmpv1.z );

            vect_mult( tmpc, &tmpv2 );
            vect_add( c, tmpc );

            free( tmpc );
        }
    }

    free( isect );

    if( c->x > 1.0f ) {
        c->x = 1.0f;
    }
    else if( c->x < 0.0f ) {
        c->x = 0.0f;
    }

    if( c->y > 1.0f ) {
        c->y = 1.0f;
    }
    else if( c->y < 0.0f ) {
        c->y = 0.0f;
    }

    if( c->z > 1.0f ) {
        c->z = 1.0f;
    }
    else if( c->z < 0.0f ) {
        c->z = 0.0f;
    }

    return c;
}

float calc_shade( primitive *light, vector *isect_pt, vector *lv, vector *ln,
        primitive *scene, int shadows ) {
    float shade = 1.0f;
    float l_dist;
    int i;
    ray r;
    vector o, dir;
    intersection *isect;
    primitive *iter;

    if( light->is_light == AREA_LIGHT && shadows > 1 && light->grid != NULL ) {
        for( i = 0; i < shadows; i++ ) {
            dir.x = light->grid[(i&63)*3] +
                ((float)rand()/RAND_MAX)*light->dx;
            dir.y = light->grid[(i&63)*3+1] +
                ((float)rand()/RAND_MAX)*light->dy;
            dir.z = light->grid[(i&63)*3+2] +
                ((float)rand()/RAND_MAX)*light->dz;

            vect_sub( &dir, isect_pt );
            l_dist = vect_length( &dir );
            vect_multf( &dir, 1.0f / l_dist );

            vect_copy( &o, &dir );
            vect_multf( &o, EPSILON );
            vect_add( &o, isect_pt );

            r.origin = &o;
            r.dir = &dir;

            for( iter = scene; iter != NULL; iter = iter->next ) {
                isect = iter->intersect( iter, &r );

                if( isect != NULL &&
                        !iter->is_light && isect->dist < l_dist ) {
                    shade -= 1.0f / shadows;
                    free( isect );
                    break;
                }

                free( isect );
            }
        }
    }
    else {
        /* setup the ray */
        vect_copy( &o, ln );
        vect_multf( &o, EPSILON );
        vect_add( &o, isect_pt );

        r.origin = &o;
        r.dir = ln;

        l_dist = vect_length( lv );

        for( iter = scene; iter != NULL; iter = iter->next ) {
            isect = iter->intersect( iter, &r );

            if( isect != NULL && !iter->is_light && isect->dist < l_dist ) {
                shade = 0.0f;
                free( isect );
                break;
            }

            free( isect );
        }
    }

    return shade;
}

intersection* intersect( ray *r, primitive *scene ) {
    intersection *isect = NULL;
    intersection *closest = NULL;
    primitive *p = scene;

    while( p != NULL ) {
        if( (isect = p->intersect( p, r )) ) {
            if( closest == NULL || isect->dist < closest->dist ) {
                free( closest );
                closest = isect;
            }
            else {
                free( isect );
            }
        }

        p = p->next;
    }

    return closest;
}

void usage( void ) {
    printf("crt usage: crt [options] <scene> <output>\n" );
    printf("options:\n" );
    printf("\t--input <in plugin>   - specifies the method of scene input\n" );
    printf("\t                        defaults to 'xml.so'\n" );
    printf("\t--output <out plugin> - specifies the fild output method\n" );
    printf("\t                        defaults to 'bmp.so'\n" );
    printf("\t--aa <aa level>       - specifies the level of anti-aliasing\n");
    printf("\t                        use, defaults to '1', values should\n");
    printf("\t                        be squares (such as 4, 16, 36, etc)\n" );
}

int main( int argc, char **argv ) {
    void *dl_out = NULL;
    void *dl_in = NULL;
    char *dl_error = NULL;
    primitive *scene = NULL;
    primitive *tmp = NULL;

    int aa_level = 1;
    int shadows = 1;
    char *input = NULL;
    char *output = NULL;

    int (*so_write)(int,int,color***,char*);
    primitive* (*so_load)(char*);

    int width = 512;
    int height = 512;

    color ***img = NULL;
    int i,j;

    printf( "CRT - version 0.2.0\n" );

    if( argc < 3 ) {
        usage();
        exit( 1 );
    }

    for( i = 1; i < argc - 2; i++ ) {
        if( strcmp( argv[i], "--help" ) == 0 ) {
            usage();
            exit( 0 );
        }
        else if( strcmp( argv[i], "--input" ) == 0 ) {
            if( i == argc - 3 || (strncmp( argv[i+1], "--", 2 ) == 0) ) {
                fprintf( stderr, "*** crt: missing argument to \"--input\"\n" );
                exit( 1 );
            }
            else {
                input = argv[i+1];
            }
        }
        else if( strcmp( argv[i], "--output" ) == 0 ) {
            if( i == argc - 3 || (strncmp( argv[i+1], "--", 2 ) == 0) ) {
                fprintf( stderr,"*** crt: missing argument to \"--output\"\n" );
                exit( 1 );
            }
            else {
                output = argv[i+1];
            }
        }
        else if( strcmp( argv[i], "--aa" ) == 0 ) {
            if( i == argc - 3 || (strncmp( argv[i+1], "--", 2 ) == 0) ) {
                fprintf( stderr, "*** crt: missing argument to \"--aa\"\n" );
                exit( 1 );
            }
            else {
                aa_level = atoi( argv[i+1] );
            }
        }
        else if( strcmp( argv[i], "--shadows" ) == 0 ) {
            if( i == argc - 3 || (strncmp( argv[i+1], "--", 2 ) == 0) ) {
                fprintf(stderr, "*** crt: missing argument to \"--shadow\"\n");
                exit( 1 );
            }
            else {
                shadows = atoi( argv[i+1] );
            }
        }
        else if( strcmp( argv[i], "--width" ) == 0 ) {
            if( i == argc - 3 || (strncmp( argv[i+1], "--", 2 ) == 0) ) {
                fprintf( stderr, "*** crt: missing argument to \"--width\"\n" );
            }
            else {
                width = atoi( argv[i+1] );
            }
        }
        else if( strcmp( argv[i], "--height" ) == 0 ) {
            if( i == argc - 3 || (strncmp( argv[i+1], "--", 2 ) == 0) ) {
                fprintf( stderr,"*** crt: missing argument to \"--height\"\n" );
            }
            else {
                height = atoi( argv[i+1] );
            }
        }
    }

    /* setup the output plugin */
    if( output == NULL ) {
        dl_out = dlopen( "./outputs/bmp.so", RTLD_LAZY );
    }
    else {
        dl_out = dlopen( output, RTLD_LAZY );
    }
    if( !dl_out ) {
        fprintf( stderr, "*** error: %s\n", dlerror() );
        exit( 1 );
    }

    dlerror();      /* clear error code */
    so_write = (int (*)(int,int,color***,char*))dlsym( dl_out, "write_img" );
    if( (dl_error = dlerror()) != NULL ) {
        fprintf( stderr, "*** error: %s\n", dl_error );
        exit( 1 );
    }

    /* setup the input plugin */
    if( input == NULL ) {
        dl_in = dlopen( "./inputs/xml.so", RTLD_LAZY );
    }
    else {
        dl_in = dlopen( input, RTLD_LAZY );
    }
    if( !dl_in ) {
        fprintf( stderr, "*** error: %s\n", dlerror() );
        exit( 1 );
    }
    
    dlerror();
    so_load = (primitive* (*)(char*))dlsym( dl_in, "load_scene" );
    if( (dl_error = dlerror()) != NULL ) {
        fprintf( stderr, "*** error: %s\n", dl_error );
        exit( 1 );
    }

    /* setup the image's memory */
    img = malloc( height * sizeof(color**) );
    if( img == NULL ) {
        fprintf( stderr, "*** error: could not allocate image memory\n" );
        exit( 1 );
    }

    for( i = 0; i < height; i++ ) {
        img[i] = calloc( width, sizeof( color* ) );
        if( img[i] == NULL ) {
            fprintf( stderr, "*** error: could not allocate image memory\n" );
            exit( 1 );
        }
    }

    scene = so_load( argv[argc-2] );

    render( width, height, img, scene, aa_level, shadows );

    so_write( width, height, img, argv[argc-1] );

    for( i = 0; i < height; i++ ) {
        for( j = 0; j < width; j++ ) {
            /* free the color's memory */
            free( img[i][j] );
        }

        /* free the line's memory */
        free( img[i] );
    }

    /* free the image memory */
    free( img );

    /* free the scene object's memory */

    while( scene != NULL ) {
        free( scene->data );
        if( scene->grid != NULL ) {
            free( scene->grid );
        }

        tmp = scene->next;
        free( scene );

        scene = tmp;
    }

    dlclose( dl_out );
    dlclose( dl_in );

    return 0;
}


vector.c
#include <math.h>
#include "vector.h"

void vect_add( vector *a, vector *b ) {
    a->x += b->x;
    a->y += b->y;
    a->z += b->z;
}

void vect_addf( vector *a, float f ) {
    a->x += f;
    a->y += f;
    a->z += f;
}

void vect_sub( vector *a, vector *b ) {
    a->x -= b->x;
    a->y -= b->y;
    a->z -= b->z;
}

void vect_subf( vector *a, float f ) {
    a->x -= f;
    a->y -= f;
    a->z -= f;
}

void vect_mult( vector *a, vector *b ) {
    a->x *= b->x;
    a->y *= b->y;
    a->z *= b->z;
}

void vect_multf( vector *a, float f ) {
    a->x *= f;
    a->y *= f;
    a->z *= f;
}

float vect_length( vector *v ) {
    return sqrtf( v->x*v->x + v->y*v->y + v->z*v->z );
}

void vect_normalize( vector *v ) {
    float length = 0.0f;
    float tmp = 0.0f;

    length = sqrtf( v->x * v->x + v->y * v->y + v->z * v->z );
    tmp = 1.0f / length;

    v->x *= tmp;
    v->y *= tmp;
    v->z *= tmp;
}

float vect_dot( vector *a, vector *b ) {
    return( a->x * b->x + a->y * b->y + a->z * b->z );
}

inline void vect_copy( vector *dest, vector *src ) {
    dest->x = src->x;
    dest->y = src->y;
    dest->z = src->z;
}


vector.h
#ifndef __VECTOR__H__
#define __VECTOR__H__

typedef struct {
    float x;
    float y;
    float z;
} vector;

typedef struct {
    vector *origin;
    vector *dir;
} ray;

typedef vector color;

/*
 * vect_add - computes a = a + b
 */
void vect_add( vector *a, vector *b );

/*
 * vect_addf - computes a = a + f
 */
void vect_addf( vector *a, float f );

/*
 * vect_sub - computes a = a - b
 */
void vect_sub( vector *a, vector *b );

/*
 * vect_subf - computes a = a - f
 */
void vect_subf( vector *a, float f );

/*
 * vect_mult - computes a = a * b
 */
void vect_mult( vector *a, vector *b );

/*
 * vect_multf -computes a = a * f
 */
void vect_multf( vector *a, float f );

/*
 * vect_length - returns the length of the vector
 */
float vect_length( vector *v );

/*
 * vect_normalize - normalizes a vector
 */
void vect_normalize( vector *v );

/*
 * vect_dot - computes the dot product of a and b
 */
float vect_dot( vector *a, vector *b );

/*
 * vect_copy - copies the values from one vector
 * to another
 */
inline void vect_copy( vector *dest, vector *src );

#endif


objects.h
#ifndef __OBJECTS__H__
#define __OBJECTS__H__

#define SPHERE  0
#define BOX     1
#define PLANE   2

#define NOT_LIGHT       0
#define POINT_LIGHT     1
#define AREA_LIGHT      2

#include "vector.h"

struct _isect;
typedef struct _isect intersection;
typedef struct _prim primitive;

typedef struct {
    color col;
    float diffuse, refl, refr, specular, absorb;
    int is_refr;
} material;

struct _prim {
    primitive *next;

    char type;
    char is_light;

    vector center;

    material mat;

    void *data;

    float *grid;
    float dx;
    float dy;
    float dz;

    intersection* (*intersect)( primitive*, ray* );
    void (*normal)( primitive*, vector*, vector* );
};

struct _isect {
    primitive *prim;
    float dist;
    char inside;
};

typedef struct {
    vector point;
    vector size;
} box_data;

typedef struct {
    vector center;
    float radius;
} sphere_data;

typedef struct {
    vector normal;
    float dist;
} plane_data;

intersection* box_isect( primitive *prim, ray *r );
void box_normal( primitive *box, vector *isect, vector *ret );
intersection* sphere_isect( primitive *prim, ray *r );
void sphere_normal( primitive *sphere, vector *isect, vector *ret );
intersection* plane_isect( primitive *prim, ray *r );
void plane_normal( primitive *plane, vector *isect, vector *ret );

#endif


inputs/xml.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <expat.h>

#include "../objects.h"
#include "../vector.h"

#define BUF_SIZE    4096

typedef struct {
    XML_Parser parser;
    primitive *prim;
    primitive *head;
} parse_data;

static void parse_xyz( float*, const char** );
static void parse_light( primitive*, const char** );
static float parse_attrf( const char*, const char**, float );
static const char* parse_attrs( const char*, const char** );

static void parse_xyz( float *xyz, const char **attr ) {
    int i;

    for( i = 0; attr[i]; i += 2 ) {
        if( strcmp( attr[i], "x" ) == 0 ) {
            xyz[0] = (float)atof( attr[i+1] );
        }
        else if( strcmp( attr[i], "y" ) == 0 ) {
            xyz[1] = (float)atof( attr[i+1] );
        }
        else if( strcmp( attr[i], "z" ) == 0 ) {
            xyz[2] = (float)atof( attr[i+1] );
        }
    }
}

static void parse_light( primitive *p, const char **attr ) {
    const char *light = parse_attrs( "light", attr );
    int i;
    float y,z;

    if( light == NULL ) {
        p->is_light = NOT_LIGHT;
    }
    else if( strcmp( light, "point" ) == 0 ) {
        p->is_light = POINT_LIGHT;
    }
    else {
        p->is_light = AREA_LIGHT;

        if( p->type == BOX ) {
            p->grid = (float*)malloc( 192 * sizeof( float ) );

            p->dx = ((box_data*)p->data)->size.x * 0.25f;
            p->dy = ((box_data*)p->data)->size.y * 0.25f;
            p->dz = ((box_data*)p->data)->size.z * 0.25f;

            for( i = 0; i < 64; i++ ) {
                p->grid[ i*3 ] = ((box_data*)p->data)->point.x + (i%4)*p->dx;

                if( i % 4 == 0 ) {
                    z = ((box_data*)p->data)->point.z + ((float)(i>=16?i%16:i)/4.0f)*p->dz;
                    if( i % 16 == 0 ) {
                        y = ((box_data*)p->data)->point.y + ((float)i/16.0f)*p->dy;
                    }
                }

                p->grid[ i*3+1 ] = y;
                p->grid[ i*3+2 ] = z;
            }

            /*
            for( i = 0; i < 16; i++ ) {
                p->grid[i*2] = ((box_data*)p->data)->point.x + (i%4)*p->dx;
                if( i % 4 == 0 ) {
                    z = ((box_data*)p->data)->point.z + ((float)i/4.0f)*p->dz;
                }

                p->grid[i*2+1] = z;
            }
            */
        }
    }
}


static float parse_attrf( const char *name, const char **attr, float def ) {
    int i;

    for( i = 0; attr[i]; i += 2 ) {
        if( strcmp( attr[i], name ) == 0 ) {
            return( (float)atof( attr[i+1] ) );
        }
    }

    return def;
}

static const char* parse_attrs( const char *name, const char **attr ) {
    int i;

    for( i = 0; attr[i]; i += 2 ) {
        if( strcmp( attr[i], name ) == 0 ) {
            return( attr[i+1] );
        }
    }

    return NULL;
}

static char parse_attrb( const char *name, const char **attr, int def ) {
    const char *truefalse = parse_attrs( name, attr );

    if( truefalse != NULL && strcmp( truefalse, "true" ) == 0 ) {
        return 1;
    }
    else {
        return 0;
    }

    return def;
}

void start( void *data, const char *el, const char **attr ) {
    parse_data *pdata = (parse_data*)data;
    float xyz[3];

    if( pdata == NULL ) {
        fprintf( stderr, "*** xml: parse data is NULL\n" );
        exit( 1 );
    }

    if( strcmp( el, "scene" ) == 0 ) {
    }
    else if( pdata->prim == NULL ) {
        pdata->prim = (primitive*)calloc( 1, sizeof( primitive ) );

        if( strcmp( el, "sphere" ) == 0 ) {
            pdata->prim->type = SPHERE;

            pdata->prim->intersect = &sphere_isect;
            pdata->prim->normal = &sphere_normal;

            pdata->prim->data = (sphere_data*)malloc( sizeof( sphere_data ) );

            parse_light( pdata->prim, attr );
            parse_xyz( xyz, attr );
            ((sphere_data*)pdata->prim->data)->radius = parse_attrf(
                "radius", attr, 0.0f );

            ((sphere_data*)pdata->prim->data)->center.x = xyz[0];
            ((sphere_data*)pdata->prim->data)->center.y = xyz[1];
            ((sphere_data*)pdata->prim->data)->center.z = xyz[2];

            vect_copy( &pdata->prim->center,
                    &((sphere_data*)pdata->prim->data)->center );
            
            /* set some defaults */
            pdata->prim->mat.col.x = 1.0f;
            pdata->prim->mat.col.y = 1.0f;
            pdata->prim->mat.col.z = 1.0f;

            pdata->prim->mat.diffuse = 1.0f;
            pdata->prim->mat.specular = 0.0f;
            pdata->prim->mat.refl = 0.0f;
            pdata->prim->mat.is_refr = 0;
        }
        else if( strcmp( el, "plane" ) == 0 ) {
            pdata->prim->type = PLANE;

            pdata->prim->intersect = &plane_isect;
            pdata->prim->normal = &plane_normal;

            pdata->prim->data = (plane_data*)malloc( sizeof( plane_data ) );

            parse_light( pdata->prim, attr );
            parse_xyz( xyz, attr );

            ((plane_data*)pdata->prim->data)->normal.x = xyz[0];
            ((plane_data*)pdata->prim->data)->normal.y = xyz[1];
            ((plane_data*)pdata->prim->data)->normal.z = xyz[2];
            ((plane_data*)pdata->prim->data)->dist = parse_attrf(
                "distance", attr, 0.0f );

            vect_copy( &pdata->prim->center,
                    &((plane_data*)pdata->prim->data)->normal );

            /* set some defaults */
            pdata->prim->mat.col.x = 1.0f;
            pdata->prim->mat.col.y = 1.0f;
            pdata->prim->mat.col.z = 1.0f;

            pdata->prim->mat.diffuse = 1.0f;
            pdata->prim->mat.specular = 0.0f;
            pdata->prim->mat.refl = 0.0f;
            pdata->prim->mat.is_refr = 0;
        }
        else if( strcmp( el, "box" ) == 0 ) {
            pdata->prim->type = BOX;

            pdata->prim->intersect = &box_isect;
            pdata->prim->normal = &box_normal;

            pdata->prim->data = (box_data*)malloc( sizeof( box_data ) );

            parse_xyz( xyz, attr );

            ((box_data*)pdata->prim->data)->point.x = xyz[0];
            ((box_data*)pdata->prim->data)->point.y = xyz[1];
            ((box_data*)pdata->prim->data)->point.z = xyz[2];
            ((box_data*)pdata->prim->data)->size.x = parse_attrf(
                "width", attr, 0.0f );
            ((box_data*)pdata->prim->data)->size.y = parse_attrf(
                "height", attr, 0.0f );
            ((box_data*)pdata->prim->data)->size.z = parse_attrf(
                "depth", attr, 0.0f );

            pdata->prim->center.x = xyz[0] +
                (0.5f * ((box_data*)pdata->prim->data)->size.x);
            pdata->prim->center.y = xyz[1] +
                (0.5f * ((box_data*)pdata->prim->data)->size.y);
            pdata->prim->center.z = xyz[2] +
                (0.5f * ((box_data*)pdata->prim->data)->size.z);

            parse_light( pdata->prim, attr );

            /* set some defaults */
            pdata->prim->mat.col.x = 1.0f;
            pdata->prim->mat.col.y = 1.0f;
            pdata->prim->mat.col.z = 1.0f;

            pdata->prim->mat.diffuse = 1.0f;
            pdata->prim->mat.specular = 0.0f;
            pdata->prim->mat.refl = 0.0f;
            pdata->prim->mat.is_refr = 0;
        }
        else {
            fprintf( stderr,
                    "*** xml: unknown primitive type \"%s\" at line %d\n",
                    el, XML_GetCurrentLineNumber( pdata->parser ) );
            exit( 1 );
        }

        pdata->prim->next = pdata->head;
        pdata->head = pdata->prim;
    }
    else if( strcmp( el, "material" ) == 0 ) {
        pdata->prim->mat.col.x = parse_attrf( "r", attr, 1.0f );
        pdata->prim->mat.col.y = parse_attrf( "g", attr, 1.0f );
        pdata->prim->mat.col.z = parse_attrf( "b", attr, 1.0f );

        pdata->prim->mat.diffuse = parse_attrf( "diffuse", attr, 1.0f );
        pdata->prim->mat.specular = parse_attrf( "specular", attr, 0.0f );
        pdata->prim->mat.refl = parse_attrf( "reflectivity", attr, 0.0f );
        pdata->prim->mat.is_refr = parse_attrb( "refractive", attr, 0 );
        if( pdata->prim->mat.is_refr ) {
            pdata->prim->mat.refr = parse_attrf( "refractivity", attr, 1.0f );
        }
    }
    else if( strcmp( el, "light" ) == 0 ) {
    }
}

void end( void *data, const char *el ) {
    parse_data *pdata = (parse_data*)data;
    if( strcmp( el, "scene" ) == 0 ) {
    }
    else {
        pdata->prim = NULL;
    }
}

primitive* load_scene( char *file ) {
    primitive *head = NULL;
    char buff[BUF_SIZE];
    FILE *ffile;
    parse_data *pdata = (parse_data*)calloc( 1, sizeof( parse_data ) );

    if( pdata == NULL ) {
        fprintf( stderr, "*** xml: could not allocate memory for parsing\n" );
        exit( 1 );
    }

    pdata->parser = XML_ParserCreate( NULL );
    
    if( pdata->parser == NULL ) {
        fprintf( stderr, "*** xml: could not allocate memory for parser\n" );
        exit( 1 );
    }

    ffile = fopen( file, "r" );
    if( ffile == NULL ) {
        fprintf( stderr, "*** xml: could not open %s for reading\n", file );
        exit( 1 );
    }

    XML_SetElementHandler( pdata->parser, start, end );
    XML_SetUserData( pdata->parser, pdata );

    for( ;; ) {
        int done;
        int len;

        len = fread( buff, 1, BUF_SIZE, ffile );
        if( ferror( ffile ) ) {
            fprintf( stderr, "*** xml: read error\n" );
            exit( 1 );
        }

        done = feof( ffile );

        if( ! XML_Parse( pdata->parser, buff, len, done ) ) {
            fprintf( stderr, "*** xml: parse error at line %d:\n%s\n",
                    XML_GetCurrentLineNumber( pdata->parser ),
                    XML_ErrorString( XML_GetErrorCode( pdata->parser ) ) );
            exit( 1 );
        }

        if( done ) break;
    }

#if 0
    p = (primitive*)malloc( sizeof( primitive ) );

    p->type = PLANE;
    p->is_light = NOT_LIGHT;

    p->intersect = &plane_isect;
    p->normal = &plane_normal;

    p->data = (plane_data*)malloc( sizeof( plane_data ) );

    ((plane_data*)p->data)->normal.x = 0.0f;
    ((plane_data*)p->data)->normal.y = 1.0f;
    ((plane_data*)p->data)->normal.z = 0.0f;
    ((plane_data*)p->data)->dist = 2.0f;

    p->center.x = 0.0f;
    p->center.y = 0.0f;
    p->center.z = 0.0f;

    p->mat.diffuse = 1.0f;
    p->mat.specular = 0.0f;
    p->mat.refl = 0.0f;
    p->mat.is_refr = 0;

    p->mat.col.x = 0.46667f;
    p->mat.col.y = 0.52941f;
    p->mat.col.z = 0.60784f;

    p->next = head;
    head = p;
#endif

    fclose( ffile );
    XML_ParserFree( pdata->parser );

    head = pdata->head;
    free( pdata );

    return head;
}
